### ISSUE TITLE
Support nested directories in custom dataset training

### ISSUE THREAD

**Issue by priteshgohil** [ultralytics/yolov5#1379]
## 🚀 Feature
Support nested folder structure to create dataset while training or testing model. Currently, dataset.py is finding files only at the top layer of the folder.

## Motivation
Data are still separated in train/test folder. However, adding a feature to [datasets.py](https://github.com/ultralytics/yolov5/blob/master/utils/datasets.py) to walk in nested folder structure will help us to train network without much modification on the dataset directories. 

I'm getting error because of the following line,
https://github.com/ultralytics/yolov5/blob/04081f810270712ba3a69577c47e5dcfa850fa90/utils/datasets.py#L357

eg. My image folder looks like
```
train
|___cat
    |____image1.png
    |____image2.png
    |____image3.png
|___dog
```
and the same structure in the test folder.

**Comment by priteshgohil** [comment]
Currently, I'm using the following fix and its working

![image](https://user-images.githubusercontent.com/43172056/98993186-71a08280-252e-11eb-9d59-6290b9a07403.png)

**Comment by glenn-jocher** [comment]
@priteshgohil that's interesting, thanks for the feedback! Do you have labels in parallel nested folders?

**Comment by glenn-jocher** [comment]
@priteshgohil try PR #1380

**Comment by priteshgohil** [comment]
Hi @glenn-jocher, Yes labels follow the similar folder structure to image.  I tried the PR  and on debug found that code hangs on at `f += glob.glob(str(p) + '/**/*.*', recursive=True)`.

Can you check above line of code directly by setting up a small set of directories and files in it?

**Comment by glenn-jocher** [comment]
@priteshgohil I already tested it by placing the coco128 images inside the the /images folder. Works great, sees all 256 images.

**Comment by glenn-jocher** [comment]
@priteshgohil not sure why it hangs on your system. Can you reproduce the hanging in a Colab notebook?

**Comment by priteshgohil** [comment]
@glenn-jocher Sorry for the delay. I was working on something else. 
I tested the repo today and it seems to be working. I use Linux. I can't figure out the exact cause but it must be something from my side.  I tried today to reproduce the hanging issue in terminal, but it was flawless this time (I am surprised why last time failed working in training and even in the terminal when I tested)
Thanks :)
